File: ios/SalmonTrollingAssistant/SalmonTrollingAssistant/Domain/Services/CatchAnalyticsServiceImpl.swift
import Foundation
import Combine

class CatchAnalyticsServiceImpl: CatchAnalyticsService {
    private let catchLoggingService: CatchLoggingService
    
    init(catchLoggingService: CatchLoggingService) {
        self.catchLoggingService = catchLoggingService
    }
    
    var catchCountBySpecies: AnyPublisher<[FishSpecies: Int], Never> {
        return catchLoggingService.catchHistory
            .map { catches in
                Dictionary(grouping: catches, by: { $0.species })
                    .mapValues { $0.count }
            }
            .eraseToAnyPublisher()
    }
    
    var catchCountByLocation: AnyPublisher<[String: Int], Never> {
        return catchLoggingService.catchHistory
            .map { catches in
                Dictionary(grouping: catches, by: { $0.locationId })
                    .mapValues { $0.count }
            }
            .eraseToAnyPublisher()
    }
    
    var catchCountByMonth: AnyPublisher<[Int: Int], Never> {
        return catchLoggingService.catchHistory
            .map { catches in
                let calendar = Calendar.current
                let groupedByMonth = Dictionary(grouping: catches) { catch in
                    calendar.component(.month, from: catch.timestamp) - 1 // 0-based month
                }
                return groupedByMonth.mapValues { $0.count }
            }
            .eraseToAnyPublisher()
    }
    
    var averageSizeBySpecies: AnyPublisher<[FishSpecies: Double], Never> {
        return catchLoggingService.catchHistory
            .map { catches in
                let catchesWithSize = catches.filter { $0.size != nil }
                let groupedBySpecies = Dictionary(grouping: catchesWithSize, by: { $0.species })
                
                return groupedBySpecies.mapValues { catches in
                    let sizes = catches.compactMap { $0.size }
                    return sizes.isEmpty ? 0 : sizes.reduce(0, +) / Double(sizes.count)
                }
            }
            .eraseToAnyPublisher()
    }
    
    var averageWeightBySpecies: AnyPublisher<[FishSpecies: Double], Never> {
        return catchLoggingService.catchHistory
            .map { catches in
                let catchesWithWeight = catches.filter { $0.weight != nil }
                let groupedBySpecies = Dictionary(grouping: catchesWithWeight, by: { $0.species })
                
                return groupedBySpecies.mapValues { catches in
                    let weights = catches.compactMap { $0.weight }
                    return weights.isEmpty ? 0 : weights.reduce(0, +) / Double(weights.count)
                }
            }
            .eraseToAnyPublisher()
    }
    
    var mostSuccessfulEquipment: AnyPublisher<[(String, Int)], Never> {
        return catchLoggingService.catchHistory
            .map { catches in
                // Flatten the list of equipment used in each catch
                let allEquipment = catches.flatMap { $0.equipmentUsed }
                
                // Count occurrences of each equipment
                var equipmentCounts: [String: Int] = [:]
                for equipment in allEquipment {
                    equipmentCounts[equipment, default: 0] += 1
                }
                
                // Sort by count in descending order
                return equipmentCounts.sorted { $0.value > $1.value }
            }
            .eraseToAnyPublisher()
    }
    
    var mostSuccessfulLocations: AnyPublisher<[(String, Int)], Never> {
        return catchLoggingService.catchHistory
            .map { catches in
                let locationCounts = Dictionary(grouping: catches, by: { $0.locationId })
                    .mapValues { $0.count }
                
                return locationCounts.sorted { $0.value > $1.value }
            }
            .eraseToAnyPublisher()
    }
    
    var catchTrendOverTime: AnyPublisher<[(Date, Int)], Never> {
        return catchLoggingService.catchHistory
            .map { catches in
                let calendar = Calendar.current
                
                // Group catches by day
                let catchesByDay = Dictionary(grouping: catches) { catch in
                    calendar.startOfDay(for: catch.timestamp)
                }
                
                // Convert to array of (date, count) pairs and sort by date
                return catchesByDay.map { (date, catches) in
                    (date, catches.count)
                }.sorted { $0.0 < $1.0 }
            }
            .eraseToAnyPublisher()
    }
    
    var personalizedRecommendations: AnyPublisher<[String], Never> {
        return catchLoggingService.catchHistory
            .map { catches in
                if catches.isEmpty {
                    return ["Start logging your catches to get personalized recommendations!"]
                }
                
                var recommendations: [String] = []
                
                // Most successful species
                let speciesCounts = Dictionary(grouping: catches, by: { $0.species })
                    .mapValues { $0.count }
                
                if let mostCaughtSpecies = speciesCounts.max(by: { $0.value < $1.value })?.key {
                    recommendations.append("You've had the most success catching \(mostCaughtSpecies.rawValue.capitalized).")
                }
                
                // Most successful location
                let locationCounts = Dictionary(grouping: catches, by: { $0.locationId })
                    .mapValues { $0.count }
                
                if let mostSuccessfulLocation = locationCounts.max(by: { $0.value < $1.value })?.key {
                    recommendations.append("Your most productive fishing spot is location \(mostSuccessfulLocation).")
                }
                
                // Most successful equipment
                let allEquipment = catches.flatMap { $0.equipmentUsed }
                var equipmentCounts: [String: Int] = [:]
                for equipment in allEquipment {
                    equipmentCounts[equipment, default: 0] += 1
                }
                
                let mostSuccessfulEquipment = equipmentCounts.sorted { $0.value > $1.value }
                    .prefix(3)
                    .map { $0.key }
                
                if !mostSuccessfulEquipment.isEmpty {
                    recommendations.append("Your most effective equipment includes: \(mostSuccessfulEquipment.joined(separator: ", ")).")
                }
                
                // Best time of year
                let calendar = Calendar.current
                let monthCounts = Dictionary(grouping: catches) { catch in
                    calendar.component(.month, from: catch.timestamp) - 1 // 0-based month
                }.mapValues { $0.count }
                
                if let bestMonth = monthCounts.max(by: { $0.value < $1.value })?.key {
                    let dateFormatter = DateFormatter()
                    dateFormatter.dateFormat = "MMMM"
                    let date = calendar.date(from: DateComponents(month: bestMonth + 1))!
                    let monthName = dateFormatter.string(from: date)
                    
                    recommendations.append("Your best fishing month appears to be \(monthName).")
                }
                
                return recommendations
            }
            .eraseToAnyPublisher()
    }
}

File: android/SalmonTrollingAssistant/app/src/main/java/com/example/salmontrollingassistant/data/service/CatchAnalyticsServiceImpl.kt
package com.example.salmontrollingassistant.data.service

import com.example.salmontrollingassistant.domain.model.CatchData
import com.example.salmontrollingassistant.domain.model.FishSpecies
import com.example.salmontrollingassistant.domain.service.CatchAnalyticsService
import com.example.salmontrollingassistant.domain.service.CatchLoggingService
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import java.util.Calendar
import java.util.Date
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CatchAnalyticsServiceImpl @Inject constructor(
    private val catchLoggingService: CatchLoggingService
) : CatchAnalyticsService {
    
    override fun getCatchCountBySpecies(): Flow<Map<FishSpecies, Int>> {
        return catchLoggingService.getCatchHistory().map { catches ->
            catches.groupingBy { it.species }.eachCount()
        }
    }
    
    override fun getCatchCountByLocation(): Flow<Map<String, Int>> {
        return catchLoggingService.getCatchHistory().map { catches ->
            catches.groupingBy { it.locationId }.eachCount()
        }
    }
    
    override fun getCatchCountByMonth(): Flow<Map<Int, Int>> {
        return catchLoggingService.getCatchHistory().map { catches ->
            catches.groupBy { catch ->
                val calendar = Calendar.getInstance()
                calendar.timeInMillis = catch.timestamp
                calendar.get(Calendar.MONTH)
            }.mapValues { it.value.size }
        }
    }
    
    override fun getAverageSizeBySpecies(): Flow<Map<FishSpecies, Double>> {
        return catchLoggingService.getCatchHistory().map { catches ->
            catches
                .filter { it.size != null }
                .groupBy { it.species }
                .mapValues { entry ->
                    entry.value.mapNotNull { it.size }.average()
                }
        }
    }
    
    override fun getAverageWeightBySpecies(): Flow<Map<FishSpecies, Double>> {
        return catchLoggingService.getCatchHistory().map { catches ->
            catches
                .filter { it.weight != null }
                .groupBy { it.species }
                .mapValues { entry ->
                    entry.value.mapNotNull { it.weight }.average()
                }
        }
    }
    
    override fun getMostSuccessfulEquipment(): Flow<List<Pair<String, Int>>> {
        return catchLoggingService.getCatchHistory().map { catches ->
            // Flatten the list of equipment used in each catch
            val allEquipment = catches.flatMap { it.equipmentUsed }
            
            // Count occurrences of each equipment
            val equipmentCounts = allEquipment.groupingBy { it }.eachCount()
            
            // Sort by count in descending order
            equipmentCounts.entries
                .sortedByDescending { it.value }
                .map { Pair(it.key, it.value) }
        }
    }
    
    override fun getMostSuccessfulLocations(): Flow<List<Pair<String, Int>>> {
        return catchLoggingService.getCatchHistory().map { catches ->
            catches
                .groupingBy { it.locationId }
                .eachCount()
                .entries
                .sortedByDescending { it.value }
                .map { Pair(it.key, it.value) }
        }
    }
    
    override fun getCatchTrendOverTime(): Flow<List<Pair<Date, Int>>> {
        return catchLoggingService.getCatchHistory().map { catches ->
            // Group catches by day
            val calendar = Calendar.getInstance()
            
            catches
                .groupBy { catch ->
                    calendar.timeInMillis = catch.timestamp
                    calendar.set(Calendar.HOUR_OF_DAY, 0)
                    calendar.set(Calendar.MINUTE, 0)
                    calendar.set(Calendar.SECOND, 0)
                    calendar.set(Calendar.MILLISECOND, 0)
                    Date(calendar.timeInMillis)
                }
                .entries
                .sortedBy { it.key }
                .map { Pair(it.key, it.value.size) }
        }
    }
    
    override fun getPersonalizedRecommendations(): Flow<List<String>> {
        return catchLoggingService.getCatchHistory().map { catches ->
            if (catches.isEmpty()) {
                return@map listOf("Start logging your catches to get personalized recommendations!")
            }
            
            val recommendations = mutableListOf<String>()
            
            // Most successful species
            val speciesCounts = catches.groupingBy { it.species }.eachCount()
            val mostCaughtSpecies = speciesCounts.maxByOrNull { it.value }?.key
            if (mostCaughtSpecies != null) {
                recommendations.add("You've had the most success catching ${mostCaughtSpecies.name}.")
            }
            
            // Most successful location
            val locationCounts = catches.groupingBy { it.locationId }.eachCount()
            val mostSuccessfulLocation = locationCounts.maxByOrNull { it.value }?.key
            if (mostSuccessfulLocation != null) {
                recommendations.add("Your most productive fishing spot is location $mostSuccessfulLocation.")
            }
            
            // Most successful equipment
            val allEquipment = catches.flatMap { it.equipmentUsed }
            val equipmentCounts = allEquipment.groupingBy { it }.eachCount()
            val mostSuccessfulEquipment = equipmentCounts.entries
                .sortedByDescending { it.value }
                .take(3)
                .map { it.key }
            
            if (mostSuccessfulEquipment.isNotEmpty()) {
                recommendations.add("Your most effective equipment includes: ${mostSuccessfulEquipment.joinToString(", ")}.")
            }
            
            // Best time of year
            val monthCounts = catches.groupBy { catch ->
                val calendar = Calendar.getInstance()
                calendar.timeInMillis = catch.timestamp
                calendar.get(Calendar.MONTH)
            }.mapValues { it.value.size }
            
            val bestMonth = monthCounts.maxByOrNull { it.value }?.key
            if (bestMonth != null) {
                val monthName = when (bestMonth) {
                    Calendar.JANUARY -> "January"
                    Calendar.FEBRUARY -> "February"
                    Calendar.MARCH -> "March"
                    Calendar.APRIL -> "April"
                    Calendar.MAY -> "May"
                    Calendar.JUNE -> "June"
                    Calendar.JULY -> "July"
                    Calendar.AUGUST -> "August"
                    Calendar.SEPTEMBER -> "September"
                    Calendar.OCTOBER -> "October"
                    Calendar.NOVEMBER -> "November"
                    Calendar.DECEMBER -> "December"
                    else -> "Unknown"
                }
                recommendations.add("Your best fishing month appears to be $monthName.")
            }
            
            recommendations
        }
    }
}